metomi/rose/config_editor/valuewidget/choice.py

```diff
@@ -0,0 +1,283 @@
+        ],
+        "format": [["--format"], {"action": "store", "metavar": "FORMAT"}],
+        "guess_groups": [
+            ["--guess-groups"],
+            {"action": "store_true", "default": False},
+        ],
+    }
+
+    def __init__(self, value, metadata, set_value, hook, arg_str=None):
+        super(ChoicesValueWidget, self).__init__(homogeneous=False, spacing=0)
+        self.value = value
+        self.metadata = metadata
+        self.set_value = set_value
+        self.hook = hook
+
+        self.opt_parser = metomi.rose.opt_parse.RoseOptionParser()
+        self.opt_parser.OPTIONS = self.OPTIONS
+        self.opt_parser.add_my_options(*list(self.OPTIONS.keys()))
+        opts, args = self.opt_parser.parse_args(shlex.split(arg_str))
+        self.all_group = opts.all_group
+        self.groups = []
+        if opts.choices is not None:
+            for choices in opts.choices:
+                self.groups.extend(metomi.rose.variable.array_split(choices))
+        self.should_edit = opts.editable
+        self.value_format = opts.format
+        self.should_guess_groups = opts.guess_groups
+        self.hints = list(args)
+
+        self.should_show_kinship = self._calc_should_show_kinship()
+        list_vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
+        list_vbox.show()
+        self._listview = metomi.rose.gtk.choice.ChoicesListView(
+            self._set_value_listview,
+            self._get_value_values,
+            self._handle_search,
+        )
+        self._listview.show()
+        list_frame = Gtk.Frame()
+        list_frame.show()
+        list_frame.add(self._listview)
+        list_vbox.pack_start(list_frame, expand=False, fill=False, padding=0)
+        self.pack_start(list_vbox, expand=True, fill=True)
+        tree_vbox = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
+        tree_vbox.show()
+        self._treeview = metomi.rose.gtk.choice.ChoicesTreeView(
+            self._set_value_treeview,
+            self._get_value_values,
+            self._get_available_values,
+            self._get_groups,
+            self._get_is_implicit,
+        )
+        self._treeview.show()
+        tree_frame = Gtk.Frame()
+        tree_frame.show()
+        tree_frame.add(self._treeview)
+        tree_vbox.pack_start(tree_frame, expand=True, fill=True, padding=0)
+        if self.should_edit:
+            add_widget = self._get_add_widget()
+            tree_vbox.pack_end(add_widget, expand=False, fill=False, padding=0)
+        self.pack_start(tree_vbox, expand=True, fill=True)
+        self._listview.connect("focus-in-event", self.hook.trigger_scroll)
+        self._treeview.connect("focus-in-event", self.hook.trigger_scroll)
+        self.grab_focus = lambda: self.hook.get_focus(self._listview)
+
+    def _handle_search(self, name):
+        return False
+
+    def _get_add_widget(self):
+        add_hbox = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL)
+        add_entry = Gtk.ComboBoxEntry()
```

Comment:
This line fails when https://github.com/metomi/rose/pull/2808/files#r1920339780 is fixed. 😢 

Looks like some of the GTK interfaces have changed.